File: apps/web/src/types/workflows.ts
export type WorkflowStatus = "DRAFT" | "READY" | "INVALID" | "ARCHIVED";

export type WorkflowRunStatus =
  "QUEUED" | "RUNNING" | "SUCCEEDED" | "FAILED" | "CANCELLED";

export type WorkflowStepRunStatus =
  "PENDING" | "RUNNING" | "SUCCEEDED" | "FAILED" | "SKIPPED" | "CANCELLED";

export type CompatibilityStatus = "COMPATIBLE" | "INCOMPATIBLE";

export interface WorkflowStep {
  id: string;
  workflow_id: string;
  position: number;
  operation_code: string;
  parameters: Record<string, unknown>;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface WorkflowSummary {
  id: string;
  name: string;
  description: string | null;
  status: WorkflowStatus;
  revision: number;
  step_count: number;
  enabled_step_count: number;
  last_run_at: string | null;
  last_run_status: WorkflowRunStatus | null;
  created_at: string;
  updated_at: string;
}

export interface WorkflowDetail extends WorkflowSummary {
  steps: WorkflowStep[];
}

export interface WorkflowListResponse {
  items: WorkflowSummary[];
  page: number;
  page_size: number;
  total: number;
}

export interface WorkflowRevision {
  revision: number;
}

export interface ValidationIssue {
  code: string;
  message: string;
  severity: "error" | "warning";
  step_id: string | null;
  position: number | null;
  field: string | null;
  column: string | null;
  details: Record<string, unknown>;
}

export interface SchemaColumn {
  name: string;
  dtype: string;
}

export interface ContractColumn {
  name: string;
  dtype: string;
  numeric_compatible: boolean;
}

export interface WorkflowCompatibility {
  status: CompatibilityStatus;
  compatible: boolean;
  reasons: ValidationIssue[];
}

export interface StepValidation {
  step_id: string;
  position: number;
  operation_code: string;
  enabled: boolean;
  valid: boolean;
  schema_before: SchemaColumn[];
  schema_after: SchemaColumn[];
  issues: ValidationIssue[];
}

export interface WorkflowValidation {
  valid: boolean;
  empty: boolean;
  issues: ValidationIssue[];
  steps: StepValidation[];
  contract: ContractColumn[];
  compatibility: WorkflowCompatibility | null;
  projected_schema: SchemaColumn[];
}

export interface WorkflowPreviewStep {
  step_id: string;
  position: number;
  operation_code: string;
  parameters: Record<string, unknown>;
  status: WorkflowStepRunStatus;
  impact: {
    rows_before: number;
    rows_after: number;
    columns_before: number;
    columns_after: number;
    changed_cell_count: number;
    removed_row_count: number;
    removed_column_count: number;
    affected_row_count: number;
    no_op: boolean;
  } | null;
  examples: {
    kind: "cell" | "row_removed" | "column_removed";
    row_index: number | null;
    column: string | null;
    before: unknown;
    after: unknown;
    reason: string | null;
  }[];
  warnings: string[];
  summary: string;
  duration_ms: number;
  error_code: string | null;
  error_message: string | null;
}

export interface WorkflowPreview {
  validation: WorkflowValidation;
  no_op: boolean;
  rows_before: number;
  rows_after: number;
  columns_before: number;
  columns_after: number;
  steps: WorkflowPreviewStep[];
  projected_quality: {
    before: number | null;
    after: number | null;
    delta: number | null;
    dimensions: {
      key: string;
      label: string;
      before: number | null;
      after: number | null;
      delta: number | null;
    }[];
  } | null;
  duration_ms: number;
}

export interface WorkflowSnapshot {
  workflow_id: string;
  name: string;
  revision: number;
  steps: Record<string, unknown>[];
}

export interface WorkflowStepRun {
  id: string;
  workflow_run_id: string;
  workflow_step_id: string | null;
  position: number;
  operation_code: string;
  status: WorkflowStepRunStatus;
  duration_ms: number | null;
  rows_before: number | null;
  rows_after: number | null;
  columns_before: number | null;
  columns_after: number | null;
  changed_cells: number | null;
  removed_rows: number | null;
  removed_columns: number | null;
  warning_summary: string | null;
  error_code: string | null;
  error_message_safe: string | null;
  step_snapshot: Record<string, unknown>;
}

export interface RunQualityDelta {
  before: number | null;
  after: number | null;
  delta: number | null;
}

export interface WorkflowRunSummary {
  id: string;
  workflow_id: string | null;
  workflow_name: string;
  workflow_revision: number;
  input_dataset_id: string | null;
  input_dataset_name: string | null;
  input_version_id: string | null;
  output_version_id: string | null;
  input_version_number: number | null;
  output_version_number: number | null;
  status: WorkflowRunStatus;
  step_count: number;
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number | null;
  error_code: string | null;
  error_message_safe: string | null;
  quality_before: number | null;
  quality_after: number | null;
}

export interface WorkflowRun extends WorkflowRunSummary {
  workflow_snapshot: WorkflowSnapshot | Record<string, unknown>;
  step_runs: WorkflowStepRun[];
  quality_delta: RunQualityDelta | null;
}

export interface WorkflowRunListResponse {
  items: WorkflowRunSummary[];
  page: number;
  page_size: number;
  total: number;
}

export interface RunListFilters {
  workflow_id?: string;
  dataset_id?: string;
  status?: WorkflowRunStatus;
  page?: number;
  page_size?: number;
}

export interface WorkflowInput {
  dataset_id: string;
  version_id: string;
}

File: apps/api/src/facilio/schemas/workflows.py
"""Workflow and workflow-run API schemas."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, Field

from facilio.schemas.transformations import (
    ChangeExampleData,
    QualityDeltaData,
    TransformationImpactData,
)

WorkflowStatus = Literal["DRAFT", "READY", "INVALID", "ARCHIVED"]
WorkflowRunStatus = Literal["QUEUED", "RUNNING", "SUCCEEDED", "FAILED", "CANCELLED"]
WorkflowStepRunStatus = Literal[
    "PENDING", "RUNNING", "SUCCEEDED", "FAILED", "SKIPPED", "CANCELLED"
]
CompatibilityStatus = Literal["COMPATIBLE", "INCOMPATIBLE"]


class WorkflowStepData(BaseModel):
    id: UUID
    workflow_id: UUID
    position: int
    operation_code: str
    parameters: dict[str, Any] = Field(default_factory=dict)
    enabled: bool
    created_at: datetime
    updated_at: datetime


class WorkflowSummary(BaseModel):
    id: UUID
    name: str
    description: str | None = None
    status: WorkflowStatus
    revision: int
    step_count: int
    enabled_step_count: int
    last_run_at: datetime | None = None
    last_run_status: WorkflowRunStatus | None = None
    created_at: datetime
    updated_at: datetime


class WorkflowDetail(WorkflowSummary):
    steps: list[WorkflowStepData] = Field(default_factory=list)


class WorkflowListData(BaseModel):
    items: list[WorkflowSummary]
    page: int
    page_size: int
    total: int


class WorkflowStepSeed(BaseModel):
    operation_code: str = Field(min_length=1, max_length=64)
    parameters: dict[str, Any] = Field(default_factory=dict)
    enabled: bool = True


class CreateWorkflowRequest(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    steps: list[WorkflowStepSeed] = Field(default_factory=list, max_length=50)


class PatchWorkflowRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=4000)
    expected_revision: int | None = None


class CreateStepRequest(BaseModel):
    operation_code: str = Field(min_length=1, max_length=64)
    parameters: dict[str, Any] = Field(default_factory=dict)
    enabled: bool = True
    expected_revision: int | None = None


class PatchStepRequest(BaseModel):
    parameters: dict[str, Any] | None = None
    enabled: bool | None = None
    operation_code: str | None = Field(default=None, min_length=1, max_length=64)
    expected_revision: int | None = None


class ReorderStepsRequest(BaseModel):
    step_ids: list[UUID]
    expected_revision: int | None = None


class WorkflowInputRequest(BaseModel):
    dataset_id: UUID
    version_id: UUID


class ValidationIssueData(BaseModel):
    code: str
    message: str
    severity: Literal["error", "warning"] = "error"
    step_id: UUID | None = None
    position: int | None = None
    field: str | None = None
    column: str | None = None
    details: dict[str, Any] = Field(default_factory=dict)


class SchemaColumnData(BaseModel):
    name: str
    dtype: str


class ContractColumnData(BaseModel):
    name: str
    dtype: str
    numeric_compatible: bool = False


class CompatibilityData(BaseModel):
    status: CompatibilityStatus
    compatible: bool
    reasons: list[ValidationIssueData] = Field(default_factory=list)


class StepValidationData(BaseModel):
    step_id: UUID
    position: int
    operation_code: str
    enabled: bool
    valid: bool
    schema_before: list[SchemaColumnData] = Field(default_factory=list)
    schema_after: list[SchemaColumnData] = Field(default_factory=list)
    issues: list[ValidationIssueData] = Field(default_factory=list)


class WorkflowValidationData(BaseModel):
    valid: bool
    empty: bool
    issues: list[ValidationIssueData] = Field(default_factory=list)
    steps: list[StepValidationData] = Field(default_factory=list)
    contract: list[ContractColumnData] = Field(default_factory=list)
    compatibility: CompatibilityData | None = None
    projected_schema: list[SchemaColumnData] = Field(default_factory=list)


class WorkflowPreviewStepData(BaseModel):
    step_id: UUID
    position: int
    operation_code: str
    parameters: dict[str, Any]
    status: WorkflowStepRunStatus
    impact: TransformationImpactData | None = None
    examples: list[ChangeExampleData] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    summary: str
    duration_ms: int = 0
    error_code: str | None = None
    error_message: str | None = None


class WorkflowPreviewData(BaseModel):
    validation: WorkflowValidationData
    no_op: bool
    rows_before: int
    rows_after: int
    columns_before: int
    columns_after: int
    steps: list[WorkflowPreviewStepData]
    projected_quality: QualityDeltaData | None = None
    duration_ms: int


class WorkflowSnapshotData(BaseModel):
    workflow_id: UUID
    name: str
    revision: int
    steps: list[dict[str, Any]]


class WorkflowStepRunData(BaseModel):
    id: UUID
    workflow_run_id: UUID
    workflow_step_id: UUID | None
    position: int
    operation_code: str
    status: WorkflowStepRunStatus
    duration_ms: int | None = None
    rows_before: int | None = None
    rows_after: int | None = None
    columns_before: int | None = None
    columns_after: int | None = None
    changed_cells: int | None = None
    removed_rows: int | None = None
    removed_columns: int | None = None
    warning_summary: str | None = None
    error_code: str | None = None
    error_message_safe: str | None = None
    step_snapshot: dict[str, Any] = Field(default_factory=dict)


class WorkflowRunSummary(BaseModel):
    id: UUID
    workflow_id: UUID | None
    workflow_name: str
    workflow_revision: int
    input_dataset_id: UUID | None
    input_dataset_name: str | None = None
    input_version_id: UUID | None
    output_version_id: UUID | None
    input_version_number: int | None = None
    output_version_number: int | None = None
    status: WorkflowRunStatus
    step_count: int
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = None
    error_code: str | None = None
    error_message_safe: str | None = None
    quality_before: float | None = None
    quality_after: float | None = None


class WorkflowRunDetail(WorkflowRunSummary):
    workflow_snapshot: dict[str, Any]
    step_runs: list[WorkflowStepRunData] = Field(default_factory=list)
    quality_delta: QualityDeltaData | None = None


class WorkflowRunListData(BaseModel):
    items: list[WorkflowRunSummary]
    page: int
    page_size: int
    total: int
